File: problem-4.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <map>
#include <set>
#include <cctype>

// Function to clean a word (remove punctuation, convert to lowercase)
std::string cleanWord(const std::string &word) {
    std::string cleaned;
    for (size_t i = 0; i < word.length(); ++i) {
        char ch = word[i];
        if (std::isalpha(ch))  // Keep only letters
            cleaned += std::tolower(ch);
    }
    return cleaned;
}

// Function to create the file if it does not exist
void createFileIfNotExists(const std::string &filename) {
    std::ifstream fileCheck(filename.c_str());
    if (!fileCheck) {
        std::ofstream file(filename.c_str());
        file << "It was the best of times,\n"
             << "it was the worst of times,\n"
             << "it was the age of wisdom,\n"
             << "it was the age of foolishness,\n";
        file.close();
        std::cout << "File created with sample text: " << filename << "\n";
    } else {
        std::cout << "File exists, reading from it...\n";
    }
}

int main() {
    std::string filename = "text.txt";
    
    // Check and create the file if necessary
    createFileIfNotExists(filename);

    std::ifstream file(filename.c_str());
    if (!file) {
        std::cerr << "Error opening file.\n";
        return 1;
    }

    std::map<std::string, std::set<int> > wordMap;
    std::string line;
    int lineNumber = 0;

    while (std::getline(file, line)) {
        lineNumber++;
        std::stringstream ss(line);
        std::string word;

        while (ss >> word) {
            word = cleanWord(word);
            if (!word.empty()) {
                wordMap[word].insert(lineNumber);
            }
        }
    }

    file.close();

    // Print results
    for (std::map<std::string, std::set<int> >::const_iterator pair = wordMap.begin(); pair != wordMap.end(); ++pair) {
        std::cout << pair->first << " ";
        std::set<int>::const_iterator it = pair->second.begin();
        int first = *it, last = *it;
        ++it;
        for (; it != pair->second.end(); ++it) {
            if (*it == last + 1) {
                last = *it;
            } else {
                std::cout << first;
                if (first != last) std::cout << "-" << last;
                std::cout << " ";
                first = last = *it;
            }
        }
        std::cout << first;
        if (first != last) std::cout << "-" << last;
        std::cout << "\n";
    }

    return 0;
}

File: src/problem-4.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runTextProcessing = void 0;
var fs = require("fs");
// Function to clean a word (remove punctuation, convert to lowercase)
var cleanWord = function (word) {
    return word.replace(/[^a-zA-Z]/g, "").toLowerCase();
};
// Function to create the file if it does not exist
var createFileIfNotExists = function (filename) {
    if (!fs.existsSync(filename)) {
        var sampleText = "It was the best of times,\n\n    it was the worst of times,\n\n    it was the age of wisdom,\n\n    it was the age of foolishness,\n";
        fs.writeFileSync(filename, sampleText);
        console.log("File created with sample text: ".concat(filename));
    }
    else {
        console.log("File exists, reading from it...");
    }
};
// Function to read and process the file
var processFile = function (filename) {
    var wordMap = new Map();
    var lines = fs.readFileSync(filename, "utf-8").split("\n");
    lines.forEach(function (line, index) {
        line
            .split(/\s+/)
            .map(cleanWord)
            .filter(Boolean)
            .forEach(function (word) {
            var lineNumbers = wordMap.get(word) || new Set();
            lineNumbers.add(index + 1);
            wordMap.set(word, lineNumbers);
        });
    });
    return wordMap;
};
// Function to format and display word occurrences
var displayWordOccurrences = function (wordMap) {
    wordMap.forEach(function (lineNumbers, word) {
        var sortedLines = Array.from(lineNumbers).sort(function (a, b) { return a - b; });
        var formattedRanges = [];
        var start = sortedLines[0];
        var end = start;
        for (var i = 1; i < sortedLines.length; i++) {
            if (sortedLines[i] === end + 1) {
                end = sortedLines[i];
            }
            else {
                formattedRanges.push(start === end ? "".concat(start) : "".concat(start, "-").concat(end));
                start = end = sortedLines[i];
            }
        }
        formattedRanges.push(start === end ? "".concat(start) : "".concat(start, "-").concat(end));
        console.log("".concat(word, " ").concat(formattedRanges.join(" ")));
    });
};
// Main function
var runTextProcessing = function () {
    var filename = "text.txt";
    createFileIfNotExists(filename);
    var wordMap = processFile(filename);
    displayWordOccurrences(wordMap);
};
exports.runTextProcessing = runTextProcessing;
